core/src/test/java/org/elasticsearch/index/query/TermsQueryBuilderTest.java

```diff
@@ -0,0 +1,154 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.query;
+
+import com.google.common.collect.Lists;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.TermQuery;
+import org.elasticsearch.indices.cache.query.terms.TermsLookup;
+import org.hamcrest.Matchers;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+
+import static org.hamcrest.Matchers.*;
+
+public class TermsQueryBuilderTest extends BaseQueryTestCase<TermsQueryBuilder> {
+
+    @Override
+    protected TermsQueryBuilder doCreateTestQueryBuilder() {
+        TermsQueryBuilder query;
+        String fieldName = getRandomFieldName(OBJECT_FIELD_NAME);
+        // terms query or lookup query
+        if (randomBoolean()) {
+            // make between 0 and 5 different values of the same type
+            Object[] values = new Object[randomInt(5)];
+            for (int i = 0; i < values.length; i++) {
+                values[i] = getRandomValueForFieldName(fieldName);
+            }
+            query = new TermsQueryBuilder(fieldName, values);
+        } else {
+            // right now the mock service returns us a list of strings
+            query = new TermsQueryBuilder(STRING_FIELD_NAME);
+            query.termsLookup(randomTermsLookup());
+        }
+        if (randomBoolean()) {
+            query.minimumShouldMatch(randomInt(100) + "%");
+        }
+        if (randomBoolean()) {
+            query.disableCoord(randomBoolean());
+        }
+        return query;
+    }
+
+    private TermsLookup randomTermsLookup() {
+        return new TermsLookup(
+                randomBoolean() ? randomAsciiOfLength(10) : null,
+                randomAsciiOfLength(10),
+                randomAsciiOfLength(10),
+                randomBoolean() ? randomAsciiOfLength(10) : null,
+                randomAsciiOfLength(10)
+        );
+    }
+
+    @Override
+    protected void doAssertLuceneQuery(TermsQueryBuilder queryBuilder, Query query, QueryShardContext context) throws IOException {
+        assertThat(query, instanceOf(BooleanQuery.class));
+        BooleanQuery booleanQuery = (BooleanQuery) query;
+
+        // we only do the check below for string fields (otherwise we'd have to decode the values)
+        if (!queryBuilder.fieldName().equals(STRING_FIELD_NAME)) {
```

Comment:
it would look something like this:

```java
    protected static Object convertBytesRefToValue(String fieldName, BytesRef bytesRef) {
        switch(fieldName) {
            case DOUBLE_FIELD_NAME:
                return NumericUtils.sortableLongToDouble(NumericUtils.prefixCodedToLong(bytesRef));
            case INT_FIELD_NAME:
                return NumericUtils.prefixCodedToInt(bytesRef);
            case BOOLEAN_FIELD_NAME:
                return convertBytesRefBooleanToValue(bytesRef);
            default:
                return bytesRef.utf8ToString();
        }
    }

    private static Boolean convertBytesRefBooleanToValue(BytesRef bytesRef) {
        String sValue = bytesRef.utf8ToString();
        if (sValue.length() == 0) {
            return Boolean.FALSE;
        }
        if (sValue.length() == 1 && sValue.charAt(0) == 'F') {
            return Boolean.FALSE;
        }
        if (Booleans.parseBoolean(sValue, false)) {
            return Boolean.TRUE;
        }
        return Boolean.FALSE;
    }
```

Then the fetch service would pre-generate random values based on the type. Seems a little too much I think.